src/bin/simhash-find-all.cpp

```diff
@@ -0,0 +1,204 @@
+#include <iostream>
+#include <unordered_set>
+#include <sstream>
+#include <fstream>
+
+#include <getopt.h>
+
+#include "simhash.h"
+
+void usage(int argc, char** argv)
+{
+    std::cout << "usage: " << argv[0]
+              << " --blocks BLOCKS"
+              << " --distance DISTANCE"
+              << " --input INPUT"
+              << " --output OUTPUT"
+              << std::endl
+              << std::endl;
+    std::cout << "Read simhashes from input, find all pairs within distance bits of each"
+              << std::endl
+              << "other, writing them to output. The endianness of the output is the same"
+              << std::endl
+              << "as that of the input."
+              << std::endl
+              << std::endl;
+    std::cout << "  --blocks BLOCKS        Number of bit blocks to use." << std::endl
+              << "  --distance DISTANCE    Maximum bit distances of matches" << std::endl
+              << "  --input INPUT          Path to input ('-' for stdin)" << std::endl
+              << "  --output OUTPUT        Path to output ('-' for stdout)" << std::endl; 
+}
+
+std::unordered_set<Simhash::hash_t> read_hashes(std::istream& stream)
+{
+    std::unordered_set<Simhash::hash_t> hashes;
+    Simhash::hash_t hash(0);
+    while (!stream.eof())
+    {
+        stream.read(reinterpret_cast<char*>(&hash), sizeof(hash));
+        if (!stream.fail())
+        {
+            hashes.insert(hash);
+        }
+        else
+        {
+            break;
+        }
+    }
+    return hashes;
+}
+
+void write_matches(std::ostream& stream, const Simhash::matches_t& matches)
+{
+    for (auto it = matches.begin(); it != matches.end() && !std::cout.fail(); ++it)
+    {
+        stream.write(
+            reinterpret_cast<const char*>(&(it->first)), sizeof(Simhash::hash_t));
+        stream.write(
+            reinterpret_cast<const char*>(&(it->second)), sizeof(Simhash::hash_t));
+    }
+    stream.flush();
+}
+
+int main(int argc, char **argv) {
+
+    std::string input(""), output("");
```

Comment:
The default constructor for `std::string` is an empty string. There's no need to pass a constructor argument. This just causes additional work which the compiler can't avoid.